si_unit helper and tests

kirinrastogi opened the pull request:
Hi @thegedge I wanted to know if I was on the right track with this, and if you had any suggestions before I get too deep into this. 

Do you see anything immediately wrong with this (besides styling and the fact that there are no tests)
Anything I can improve on in the logic/ general idea of this?


fixes #88

thegedge commented on lib/measured/unit_system_builder.rb:
```diff
@@ -8,12 +8,36 @@ def unit(unit_name, aliases: [], value: nil)
   end
 
+  def si_unit(unit_name, aliases: [], value: nil, range: range = 3)
+    @units.concat build_si_units(unit_name, aliases: aliases, value: value, range: range)
```
We prefer using parentheses for all function calls except for DSL-like things. Another option is `@units += ...`

thegedge commented on lib/measured/unit_system_builder.rb:
```diff
@@ -8,12 +8,36 @@ def unit(unit_name, aliases: [], value: nil)
+
+    get_negative_prefixes(range).each_with_index do |prefix, index|
+      si_units << build_si_unit(prefix + name, concat_prefixes(aliases, prefix), value * power(index, -1))
+    end
```
As you can tell, these are very similar. What I'd recommend is actually not distinguishing between positive and negative. Something like this:

```ruby
PREFIXES = [
  ...
  ['n', 'nano', -9],
  ...
  ['k', 'kilo', 3],
  ...
]

def build_si_units(name, aliases: [], value: nil)
  value ||= 1
  PREFIXES.map do |short, long, exponent|
    long_names = aliases.map { |alias| "#{alias}#{name}" }
    build_unit("#{short}#{name}", aliases: long_names, value: value * 10**exponent)
  end
end
```

The tricky part is concatenating the names in the right way. Perhaps `si_unit` should have `short:` and `long:` keyword arguments, so that `build_si_units` can concatenate the prefixes together correctly. I wouldn't want to end up with `kmeter`, but maybe we just call that part out in the README 😄

kirinrastogi replied:
Thanks! This is a great idea since the prefixes that you suggested to add are not evenly spaced (10<sup>1</sup>, 10<sup>2</sup>, 10<sup>3</sup>, 10<sup>6</sup>)

And your suggestion for the alias concatenation is great! I'll get started on this

kirinrastogi commented:
@thegedge I changed it up and got it working! I think that the code looks ugly, so I'm working on making it prettier.

I've added 1 simple test, what should I test for? The same things in ```unit_system_builder_test.rb```?

kirinrastogi commented on lib/measured/unit_system_builder.rb:
```diff
@@ -8,12 +8,56 @@ def unit(unit_name, aliases: [], value: nil)
+  def build_si_units(name, aliases: [], value: nil)
+    value = "1 #{name}" if !value
+    size, unit = value.split(' ')
+    size ||= 1
```
I'm not sure if this is the best way to do this, it seems sketchy, as it will fail if someone sets the value to be ```1000 weight ton``` as there are 2 spaces.

thegedge replied:
So I'd rather avoid doing the `split` above. If we absolutely must, it's important to share the code that deals with this, specifically [`Unit#parse_value`](https://github.com/Shopify/measured/blob/58470f5cb085384d1d77484aeb9f21e9b7de0456/lib/measured/unit.rb#L66-L77).

What I'd actually recommend though, is special casing the base unit's value, so you start off with something like
```ruby
si_units = [build_unit(name, aliases: aliases, value: value)]
```

The cool "trick" you can use after that is that all of the prefix units can then have conversion values based on the base unit! So you could do this in your loop below:

```ruby
si_units << build_unit("#{short}#{name}", aliases: long_names, value: "10e#{exp} #{name}")
```

This means you don't have to unpack value. It's also important for a special case: what if `si_unit` is the base unit of our entire system, upon which all other conversions are based? In that case `value` is `nil` and things wouldn't work (as they are now).

thegedge commented on lib/measured/unit_system_builder.rb:
```diff
@@ -8,12 +8,56 @@ def unit(unit_name, aliases: [], value: nil)
+      ['m', 'milli', -3],
+      ['c', 'centi', -2],
+      ['d', 'deci', -1],
+      ['', '', 0],
```
Clever way of taking care of the base without having to make a special case for it👌 

That being said, special casing may make things a little nicer, and I've left a comment on your recent comment to call out why.

kirinrastogi commented on lib/measured/unit_system_builder.rb:
```diff
@@ -8,12 +8,54 @@ def unit(unit_name, aliases: [], value: nil)
+  def build_si_units(name, aliases: [], value: nil)
+    size = value.split(' ')[0] unless !value
+    size ||= 1
+    si_units = []
```
@thegedge your solution is a bit confusing to me. I was under the impression that the size was not always going to be ```"10e${exp}"```, I thought size was needed to be extracted so it could be multiplied after calculating the exponent. For example, the following case:
```
unit :lb
si_unit :long_ton, value: "2240 lb"
```
Size would need to be extracted, in this case it is ```2240```.
Should this 'feature' be waived so that si_unit only works for units that have a relation of the form ```"10e#{exp}``` ?

kirinrastogi commented on lib/measured/unit_system_builder.rb:
```diff
@@ -8,12 +8,53 @@ def unit(unit_name, aliases: [], value: nil)
+  def build_si_units(name, aliases: [], value: nil)
+    size = value.split(' ')[0] unless !value
+    size ||= 1
+    si_units = [build_unit(name, aliases: aliases, value: value)]
```
@thegedge your solution is a bit confusing to me. I was under the impression that the size was not always going to be ```"10e${exp}"```, I thought size was needed to be extracted so it could be multiplied after calculating the exponent. For example, the following case:
```
unit :lb
si_unit :long_ton, value: "2240 lb"
```
Size would need to be extracted, in this case it is ```2240```.
Should this 'feature' be waived so that si_unit only works for units that have a relation of the form ```"10e#{exp}``` ?

thegedge replied:
Right. In my solution, the "identity" element of the SI-prefixed units (in your example, `long_ton`) would be built with the given value:

```ruby
si_units = [build_unit(name, aliases: aliases, value: value)]
```

In your example, the given `value` is `"2240 lb"`, and the unit class will properly deal with parsing that string. Every other unit though can be given conversion values based on `long_ton`:

```ruby
si_units << build_unit("#{short}#{name}", aliases: long_names, value: "10e#{exp} #{name}")
```

For example, for `mega` the above line would evaluate to:

```ruby
si_units << build_unit("mlong_ton", aliases: [], value: "10e6 long_ton")
```

kirinrastogi replied:
Oh I see! I was confused and making things more complicated than needed. The spacing between say ```mlong_ton``` and ```klong_ton``` has nothing to do with it's value relative to ```lb```. Thanks for the clarification and review!

kirinrastogi commented:
@thegedge I added the changes, and switched the quotes I was using from ```'``` to ```"```

thegedge commented on lib/measured/unit_system_builder.rb:
```diff
@@ -8,12 +8,51 @@ def unit(unit_name, aliases: [], value: nil)
 
   private
 
+  def self.prefixes
```
1. This actually isn't `private` (see [this](https://til.hashrocket.com/posts/ec20caa6ef-private-class-methods-in-ruby) link)
2. Let's just make this one a constant, `SI_PREFIXES`.

thegedge commented on lib/measured/unit_system_builder.rb:
```diff
@@ -8,12 +8,51 @@ def unit(unit_name, aliases: [], value: nil)
+      si_units << build_unit("#{short}#{name}", aliases: long_names, value: "#{10 ** exp} #{name}")
+    end
+    si_units
+  end
```
Looking good now 👌

thegedge commented on test/unit_system_builder_test.rb:
```diff
@@ -37,4 +37,79 @@ class Measured::UnitSystemBuilderTest < ActiveSupport::TestCase
+      end
+    end
+
+    assert_equal (1000/1), measurable.unit_system.unit_for!(:bigg).conversion_amount
```
You can just write 1000 (a `Rational` will compare against a `Fixnum` with no issue)

thegedge commented on test/unit_system_builder_test.rb:
```diff
@@ -37,4 +37,79 @@ class Measured::UnitSystemBuilderTest < ActiveSupport::TestCase
+
+  test "#si_unit generates base unit properly" do
+    measurable = ""
+    assert_nothing_raised do
```
No need to use `assert_nothing_raised` (see [here](http://www.thagomizer.com/blog/2013/06/11/assert-nothing-raised.html)).

thegedge commented on test/unit_system_builder_test.rb:
```diff
@@ -37,4 +37,79 @@ class Measured::UnitSystemBuilderTest < ActiveSupport::TestCase
+      si_unit :ft, value: "12 in", aliases: [:fts]
+    end
+
+    prefixes = Measured::UnitSystemBuilder.prefixes.map {|short, long, exp| "#{short}ft"}
```
Use an underscore when you don't need to use something (i.e., `|short, _, _|`)

thegedge commented on test/unit_system_builder_test.rb:
```diff
@@ -37,4 +37,79 @@ class Measured::UnitSystemBuilderTest < ActiveSupport::TestCase
+    assert_equal "BOLD", measurable.unit_system.unit_for!(:BOLD).name
+  end
+
+  test "#si_unit generates base unit properly" do
```
This test is getting pretty big, which usually indicates an opportunity to split it up.

First, It's always good for your test name to be as specific as possible. "properly" isn't totally clear here.

Second, I think there are two tests here. One for the case when `si_unit` is the base unit, and the other for when `si_unit`'s conversion value is the base unit. Even though they're testing the same things, it's nice to split them up. A common way to name these tests is by using the word "when" (e.g., `#foo returns expected value when given bar`)

kirinrastogi commented:
@thegedge Thanks! I'm learning so much from this.

thegedge commented on test/unit_system_builder_test.rb:
```diff
@@ -37,4 +37,76 @@ class Measured::UnitSystemBuilderTest < ActiveSupport::TestCase
+    end
+
+    prefixes = Measured::UnitSystemBuilder::SI_PREFIXES.map {|short, _, _| "#{short}ft"}
+    prefixes += ['in', 'ft']
```
Just a couple of small style nitpicks for the above two lines:

1. [Spaces around curly braces](https://github.com/bbatsov/ruby-style-guide#spaces-braces), and
2. [prefer `%w(...)` for arrays of strings](https://github.com/bbatsov/ruby-style-guide#percent-w).

kirinrastogi commented:
@thegedge Thanks! I wasn't sure if you guys used an internal guide or a community guide when I made this, but I have fixed what you pointed out.

thegedge commented:
We roughly follow the community style guide I linked to. There's nothing to enforce it at the moment, so I'll likely just keep an eye out during code review. It's a "soft" enforcement at the moment.

>  I also fixed the bracket spacing in `def build_si_units`

Thanks :)

thegedge commented:
After we make weight / length more comprehensive, I'll put together a 2.1 release that includes this change too. We appreciate your contribution!
